packages/vats/src/orchestration.js

```diff
@@ -0,0 +1,268 @@
+      },
+      connectionHandler: {
+        /**
+         * @param {Connection} connection
+         * @param {string} localAddr
+         * @param {string} remoteAddr
+         * @param {ConnectionHandler} _connectionHandler
+         */
+        async onOpen(connection, localAddr, remoteAddr, _connectionHandler) {
+          trace(`ICA Channel Opened for ${localAddr} at ${remoteAddr}`);
+          this.state.connection = connection;
+          this.state.remoteAddress = remoteAddr;
+          this.state.localAddress = localAddr;
+          // XXX parseAddress currently throws, should it return '' instead?
+          this.state.accountAddress = parseAddress(remoteAddr);
+        },
+        async onClose(_connection, reason) {
+          trace(`ICA Channel closed. Reason: ${reason}`);
+          // XXX handle connection closing
+        },
+        async onReceive(connection, bytes) {
+          trace(`ICA Channel onReceive`, connection, bytes);
+          return '';
+        },
+      },
+    },
+  );
+
+export const OrchestrationI = M.interface('Orchestration', {
+  createAccount: M.callWhen(M.string(), M.string())
+    .optional(M.remotable('Port'))
+    .returns(M.remotable('ChainAccount')),
+  getChain: M.callWhen(M.string()).returns(M.remotable('Chain')),
+});
+
+/**
+ * @param {import('@agoric/base-zone').Zone} zone
+ * @param {ReturnType<typeof prepareChainAccount>} createChainAccount
+ */
+const prepareOrchestration = (zone, createChainAccount) =>
+  zone.exoClassKit(
+    'Orchestration',
+    {
+      admin: M.interface('OrchestrationAdmin', {
+        placeholder: M.call().returns(M.string()),
+      }),
+      public: OrchestrationI,
+    },
+    /** @param {Partial<OrchestrationPowers>} [initialPowers] */
+    initialPowers => {
+      /** @type {PowerStore} */
+      const powers = zone.detached().mapStore('PowerStore');
+      if (initialPowers) {
+        for (const [name, power] of Object.entries(initialPowers)) {
+          powers.init(/** @type {keyof OrchestrationPowers} */ (name), power);
+        }
+      }
+      return { powers, icaControllerNonce: 0 };
+    },
+    {
+      admin: {
+        placeholder() {
+          return 'test';
+        },
+      },
+      public: {
+        /**
+         * @param {import('@agoric/orchestration').ConnectionId} hostConnectionId
+         *   the counterparty connection_id
+         * @param {import('@agoric/orchestration').ConnectionId} controllerConnectionId
+         *   self connection_id
+         * @param {Port} [currentPort] if a port is provided, it will be reused
+         *   to create the account
+         * @returns {Promise<ChainAccount>}
+         */
+        async createAccount(
+          hostConnectionId,
+          controllerConnectionId,
+          currentPort,
+        ) {
+          await null;
+
+          let port;
+          if (!matches(currentPort, M.remotable('Port'))) {
+            if (currentPort) trace('Invalid Port provided, binding a new one.');
```

Comment:
if they passed a non-port for `currentPort`, shouldn't that be an error?

I think you want the check to be whether `currentPort === undefined`, which indicates whether the caller passed an argument. then validate it's actually a port and error if it isn't.

Reply:
Correct, agree this should be an error.

Aside - it seems I was mistaken in some of my understanding here. This test is failing on the second `createAccount` call:

```js
  const orchestration = await EV.vat('bootstrap').consumeItem('orchestration');

  const account1 = await EV(orchestration).createAccount(
    'connection-0',
    'connection-0',
  );
  t.truthy(account1, 'createAccount returns an account');

  const port = await EV(account1).getPort();
  t.truthy(port, 'getPort returns a port');

  const port1Addr = await EV(port).getLocalAddress();
  t.truthy(port1Addr);

  const account2 = await EV(orchestration).createAccount(
    'connection-1',
    'connection-1',
    port,
  );
  // fails -  `key channel-0:icacontroller-0 already registered in collection channelKeyToSeqAck`:
  t.truthy(account2, 'createAccount returns an account with port provided');


  const port2 = await EV(account2).getPort();
  const port2Addr = await EV(port2).getLocalAddress();
  t.is(port1Addr, port2Addr, 'port address is the same');
  ```
  
  Removing this logic for now. Port reuse is in scope for #9068